server/src/main/java/org/opensearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -407,27 +415,30 @@ protected void dispatchedShardOperationOnPrimary(
         IndexShard primary,
         ActionListener<PrimaryResult<BulkShardRequest, BulkShardResponse>> listener
     ) {
+        Span span = tracer.startSpan(SpanBuilder.from("shardPrimaryAction", clusterService.localNode().getId(), request));
```

Comment:
@rayshrey I was looking a bit deeply into how the code is organized it it looks to me this is not the right place for instrumentation: `dispatchedShardOperationOnPrimary`  / `dispatchedShardOperationOnReplica` are template methods executed from `shardOperationOnReplica` and `shardOperationOnPrimary`, I think we should instrument the later ones otherwise every single transport action that overrides  `dispatchedShardOperationOnPrimary`  / `dispatchedShardOperationOnReplica` has to repeat the same instrumentation all over again.

Reply:
@reta If we are to instrument `shardOperationOnReplica` and `shardOperationOnPrimary` in `TransportWriteAction` then we would have to add tracer to either `TransportAction`/`TransportWriteAction` (creating a `getTracer()` method in `TransportAction`/`TransportWriteAction` returning NoopTracer and overriding it in `TransdportShardBulkAction` to get the injected tracer object as done in the earlier commits)

Where do you recommend to add the `getTracer()` method - in `TranportWriteAction` or `TransportAction` ?
Or is there any other recommended way to follow here ?

Also was thinking of adding spans to indicate how much time the `shardOperationOnPrimary`/`shardOperationOnReplica` spend in the Threadpool queue before executing `dispatchedShardOperationOnPrimary`/`dispatchedShardOperationOnReplica`. So wanted to understand what would be better - to explicitly add spans in the TranportWriteAction itself or to wait for generic implementation of instrumentation in the Threadpool side ?

Reply:
> @reta If we are to instrument shardOperationOnReplica and shardOperationOnPrimary in TransportWriteAction then we would have to add tracer to either TransportAction

Correct, the `TransportWriteAction` should have tracer provided a constructor argument (and later on  `TransportAction` but this is larger change).

>  So wanted to understand what would be better - to explicitly add spans in the TranportWriteAction itself or to wait for generic implementation of instrumentation in the Threadpool side ?

I think instrumenting `Threadpool` would be ideal in this case (and we have metric for that already), however adding the span to  `TransportWriteAction` could be useful now: so we know when action was started and when it was scheduled.